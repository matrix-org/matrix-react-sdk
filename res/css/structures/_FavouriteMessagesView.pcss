/*
Copyright 2019 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.mx_FavMessagesHeader {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%; 
    flex:  0 0 50px;
    border-bottom: 1px solid $primary-hairline-color;
    background-color: $background;
    z-index: 999;
}

.mx_FavMessagesHeader_Wrapper{
    height: 44px;
    display: flex;
    align-items: center;
    min-width: 0;
    margin: 0 20px 0 16px;
    padding-top: 8px;
    border-bottom: 1px solid $system;
    justify-content: space-between;

    &--left{
        display: flex;
        align-items: center;
        flex: 0.4;
    
        & > span {
                color: $primary-content;
                font-weight: $font-semi-bold;
                font-size: $font-18px;
                margin: 0 8px;
            }
    }

    &--right{
        display: flex;
        align-items: center;
        flex: 0.6;
        justify-content: flex-end;
    }
}


.mx_FavMessagesHeader_sortButton::before {
    mask-image: url('$(res)/img/element-icons/room/sort-twoway.svg');
}

.mx_FavMessagesHeader_clearAllButton::before {
    mask-image: url('$(res)/img/element-icons/room/clear-all.svg');
}

.mx_FavMessagesHeader_cancelButton {
    background-color: $alert;
    mask: url('$(res)/img/cancel.svg');
    mask-repeat: no-repeat;
    mask-position: center;
    mask-size: 17px;
    padding: 9px;
    margin: 0 12px 0 3px;
    cursor: pointer;
}

.mx_FavMessagesHeader_Search{
    width: 70%;
}

.mx_FavouriteMessages_emptyMarker{
    display: flex;
    align-items: center;
    justify-content: center;
    font-size: 25px;
    font-weight: 600;
}

.mx_FavouriteMessages_scrollPanel {
    margin-top: 25px;
}

.mx_ClearDialog{
    width: 100%;
}