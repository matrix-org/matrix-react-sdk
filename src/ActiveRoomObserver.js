/*
Copyright 2017 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import RoomViewStore from './stores/RoomViewStore';

/**
 * Consumes changes from the RoomViewStore and notifies specific things
 * about when the active room changes. Unlike listening for RoomViewStore
 * changes, you can subscribe to only changes relevant to a particular
 * room.
 *
 * TODO: If we introduce an observer for something else, factor out
 * the adding / removing of listeners & emitting into a common class.
 */
class ActiveRoomObserver {
    constructor() {
        this._listeners = {};

        this._activeRoomId = RoomViewStore.getRoomId();
        // TODO: We could self-destruct when the last listener goes away, or at least
        // stop listening.
        this._roomStoreToken = RoomViewStore.addListener(this._onRoomViewStoreUpdate.bind(this));
    }

    addListener(roomId, listener) {
        if (!this._listeners[roomId]) this._listeners[roomId] = [];
        this._listeners[roomId].push(listener);
    }

    removeListener(roomId, listener) {
        if (this._listeners[roomId]) {
            const i = this._listeners[roomId].indexOf(listener);
            if (i > -1) {
                this._listeners[roomId].splice(i, 1);
            }
        } else {
            console.warn("Unregistering unrecognised listener (roomId=" + roomId + ")");
        }
    }

    _emit(roomId) {
        if (!this._listeners[roomId]) return;

        for (const l of this._listeners[roomId]) {
            l.call();
        }
    }

    _onRoomViewStoreUpdate() {
        // emit for the old room ID
        if (this._activeRoomId) this._emit(this._activeRoomId);

        // update our cache
        this._activeRoomId = RoomViewStore.getRoomId();

        // and emit for the new one
        if (this._activeRoomId) this._emit(this._activeRoomId);
    }
}

if (global.mx_ActiveRoomObserver === undefined) {
    global.mx_ActiveRoomObserver = new ActiveRoomObserver();
}
export default global.mx_ActiveRoomObserver;
