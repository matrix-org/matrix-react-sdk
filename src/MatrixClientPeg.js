/*
Copyright 2015, 2016 OpenMarket Ltd
Copyright 2017 Vector Creations Ltd.
Copyright 2017 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { MatrixClient, MemoryStore } from "matrix-js-sdk";

import utils from "matrix-js-sdk/lib/utils";
import EventTimeline from "matrix-js-sdk/lib/models/event-timeline";
import EventTimelineSet from "matrix-js-sdk/lib/models/event-timeline-set";
import sdk from "./index";
import createMatrixClient from "./utils/createMatrixClient";
import SettingsStore from "./settings/SettingsStore";
import MatrixActionCreators from "./actions/MatrixActionCreators";
import Modal from "./Modal";
import { verificationMethods } from "matrix-js-sdk/lib/crypto";
import MatrixClientBackedSettingsHandler from "./settings/handlers/MatrixClientBackedSettingsHandler";
import * as StorageManager from "./utils/StorageManager";
import IdentityAuthClient from "./IdentityAuthClient";

interface MatrixClientCreds {
    homeserverUrl: string;
    identityServerUrl: string;
    userId: string;
    deviceId: string;
    accessToken: string;
    guest: boolean;
}

/**
 * Wrapper object for handling the js-sdk Matrix Client object in the react-sdk
 * Handles the creation/initialisation of client objects.
 * This module provides a singleton instance of this class so the 'current'
 * Matrix Client object is available easily.
 */
class MatrixClientPeg {
    constructor() {
        this.matrixClient = null;
        this._justRegisteredUserId = null;

        // These are the default options used when when the
        // client is started in 'start'. These can be altered
        // at any time up to after the 'will_start_client'
        // event is finished processing.
        this.opts = {
            initialSyncLimit: 20
        };
        // the credentials used to init the current client object.
        // used if we tear it down & recreate it with a different store
        this._currentClientCreds = null;
    }

    /**
     * Sets the script href passed to the IndexedDB web worker
     * If set, a separate web worker will be started to run the IndexedDB
     * queries on.
     *
     * @param {string} script href to the script to be passed to the web worker
     */
    setIndexedDbWorkerScript(script) {
        createMatrixClient.indexedDbWorkerScript = script;
    }

    get(): MatrixClient {
        return this.matrixClient;
    }

    unset() {
        this.matrixClient = null;

        MatrixActionCreators.stop();
    }

    /**
     * If we've registered a user ID we set this to the ID of the
     * user we've just registered. If they then go & log in, we
     * can send them to the welcome user (obviously this doesn't
     * guarentee they'll get a chat with the welcome user).
     *
     * @param {string} uid The user ID of the user we've just registered
     */
    setJustRegisteredUserId(uid) {
        this._justRegisteredUserId = uid;
    }

    /**
     * Returns true if the current user has just been registered by this
     * client as determined by setJustRegisteredUserId()
     *
     * @returns {bool} True if user has just been registered
     */
    currentUserIsJustRegistered() {
        return (
            this.matrixClient &&
            this.matrixClient.credentials.userId === this._justRegisteredUserId
        );
    }

    /*
     * Replace this MatrixClientPeg's client with a client instance that has
     * homeserver / identity server URLs and active credentials
     */
    replaceUsingCreds(creds: MatrixClientCreds) {
        this._currentClientCreds = creds;
        this._createClient(creds);
    }

    async assign() {
        for (const dbType of ["indexeddb", "memory"]) {
            try {
                const promise = this.matrixClient.store.startup();
                console.log(
                    "MatrixClientPeg: waiting for MatrixClient store to initialise"
                );
                await promise;
                break;
            } catch (err) {
                if (dbType === "indexeddb") {
                    console.error(
                        "Error starting matrixclient store - falling back to memory store",
                        err
                    );
                    this.matrixClient.store = new MemoryStore({
                        localStorage: global.localStorage
                    });
                } else {
                    console.error("Failed to start memory store!", err);
                    throw err;
                }
            }
        }

        StorageManager.trackStores(this.matrixClient);

        // try to initialise e2e on the new client
        try {
            // check that we have a version of the js-sdk which includes initCrypto
            if (
                !SettingsStore.getValue("lowBandwidth") &&
                this.matrixClient.initCrypto
            ) {
                await this.matrixClient.initCrypto();
                StorageManager.setCryptoInitialised(true);
            }
        } catch (e) {
            if (e && e.name === "InvalidCryptoStoreError") {
                // The js-sdk found a crypto DB too new for it to use
                const CryptoStoreTooNewDialog = sdk.getComponent(
                    "views.dialogs.CryptoStoreTooNewDialog"
                );
                Modal.createDialog(CryptoStoreTooNewDialog, {
                    host: window.location.host
                });
            }
            // this can happen for a number of reasons, the most likely being
            // that the olm library was missing. It's not fatal.
            console.warn("Unable to initialise e2e", e);
        }

        const opts = utils.deepCopy(this.opts);
        // the react sdk doesn't work without this, so don't allow
        opts.pendingEventOrdering = "detached";
        opts.lazyLoadMembers = true;

        // Connect the matrix client to the dispatcher and setting handlers
        MatrixActionCreators.start(this.matrixClient);
        MatrixClientBackedSettingsHandler.matrixClient = this.matrixClient;

        return opts;
    }

    async start() {
        const opts = await this.assign();

        console.log(`MatrixClientPeg: really starting MatrixClient`);
        await this.get().startClient(opts);
        console.log(`MatrixClientPeg: MatrixClient started`);
    }

    getCredentials(): MatrixClientCreds {
        return {
            homeserverUrl: this.matrixClient.baseUrl,
            identityServerUrl: this.matrixClient.idBaseUrl,
            userId: this.matrixClient.credentials.userId,
            deviceId: this.matrixClient.getDeviceId(),
            accessToken: this.matrixClient.getAccessToken(),
            guest: this.matrixClient.isGuest()
        };
    }

    /*
     * Return the server name of the user's homeserver
     * Throws an error if unable to deduce the homeserver name
     * (eg. if the user is not logged in)
     */
    getHomeserverName() {
        const matches = /^@.+:(.+)$/.exec(this.matrixClient.credentials.userId);
        if (matches === null || matches.length < 1) {
            throw new Error("Failed to derive homeserver name from user ID!");
        }
        return matches[1];
    }

    _createClient(creds: MatrixClientCreds) {
        const opts = {
            baseUrl: creds.homeserverUrl,
            idBaseUrl: creds.identityServerUrl,
            accessToken: creds.accessToken,
            userId: creds.userId,
            deviceId: creds.deviceId,
            timelineSupport: true,
            forceTURN: !SettingsStore.getValue("webRtcAllowPeerToPeer", false),
            fallbackICEServerAllowed: !!SettingsStore.getValue(
                "fallbackICEServerAllowed"
            ),
            verificationMethods: [verificationMethods.SAS],
            unstableClientRelationAggregation: true,
            identityServer: new IdentityAuthClient()
        };

        this.matrixClient = createMatrixClient(opts);

        // we're going to add eventlisteners for each matrix event tile, so the
        // potential number of event listeners is quite high.
        this.matrixClient.setMaxListeners(500);

        this.matrixClient.setGuest(Boolean(creds.guest));

        const notifTimelineSet = new EventTimelineSet(null, {
            timelineSupport: true
        });
        // XXX: what is our initial pagination token?! it somehow needs to be synchronised with /sync.
        notifTimelineSet
            .getLiveTimeline()
            .setPaginationToken("", EventTimeline.BACKWARDS);
        this.matrixClient.setNotifTimelineSet(notifTimelineSet);
        //console.log("OPTIONS TO GET CLIENT CREDENTIALS", opts);
    }
}

if (!global.mxMatrixClientPeg) {
    global.mxMatrixClientPeg = new MatrixClientPeg();
}
export default global.mxMatrixClientPeg;
