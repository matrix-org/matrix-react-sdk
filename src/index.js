/*
Copyright 2015, 2016 OpenMarket Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

var Skinner = require('./Skinner');

var t = require('counterpart');

t.registerTranslations('en', require('../languages/en'));
t.registerTranslations('pt-BR', require('../languages/pt-BR'));

t.setFallbackLocale('en');
t.setLocale(window.navigator.languages[0]);



module.exports.loadSkin = function(skinObject) {
    Skinner.load(skinObject);
};

module.exports.resetSkin = function() {
    Skinner.reset();
};

module.exports.getComponent = function(componentName) {
    return Skinner.getComponent(componentName);
};
