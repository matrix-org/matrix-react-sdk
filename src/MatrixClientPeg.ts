/*
Copyright 2015, 2016 OpenMarket Ltd
Copyright 2017 Vector Creations Ltd.
Copyright 2017, 2018, 2019 New Vector Ltd
Copyright 2019, 2020 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { ICreateClientOpts } from 'matrix-js-sdk/src/matrix';
import {MatrixClient} from 'matrix-js-sdk/src/client';
import {MemoryStore} from 'matrix-js-sdk/src/store/memory';
import * as utils from 'matrix-js-sdk/src/utils';
import {EventTimeline} from 'matrix-js-sdk/src/models/event-timeline';
import {EventTimelineSet} from 'matrix-js-sdk/src/models/event-timeline-set';
import * as sdk from './index';
import createMatrixClient from './utils/createMatrixClient';
import SettingsStore from './settings/SettingsStore';
import MatrixActionCreators from './actions/MatrixActionCreators';
import Modal from './Modal';
import {verificationMethods} from 'matrix-js-sdk/src/crypto';
import MatrixClientBackedSettingsHandler from "./settings/handlers/MatrixClientBackedSettingsHandler";
import * as StorageManager from './utils/StorageManager';
import IdentityAuthClient from './IdentityAuthClient';
import { crossSigningCallbacks, tryToUnlockSecretStorageWithDehydrationKey } from './SecurityManager';
import {SHOW_QR_CODE_METHOD} from "matrix-js-sdk/src/crypto/verification/QRCode";
import SecurityCustomisations from "./customisations/Security";

export interface IMatrixClientCreds {
    homeserverUrl: string;
    identityServerUrl: string;
    userId: string;
    deviceId?: string;
    accessToken: string;
    guest?: boolean;
    pickleKey?: string;
    freshLogin?: boolean;
}

// TODO: Move this to the js-sdk
export interface IOpts {
    initialSyncLimit?: number;
    pendingEventOrdering?: "detached" | "chronological";
    lazyLoadMembers?: boolean;
    clientWellKnownPollPeriod?: number;
}

export interface IMatrixClientPeg {
    opts: IOpts;

    /**
     * Sets the script href passed to the IndexedDB web worker
     * If set, a separate web worker will be started to run the IndexedDB
     * queries on.
     *
     * @param {string} script href to the script to be passed to the web worker
     */
    setIndexedDbWorkerScript(script: string): void;

    /**
     * Return the server name of the user's homeserver
     * Throws an error if unable to deduce the homeserver name
     * (eg. if the user is not logged in)
     *
     * @returns {string} The homeserver name, if present.
     */
    getHomeserverName(): string;

    get(): MatrixClient;
    unset(): void;
    assign(): Promise<any>;
    start(): Promise<any>;

    getCredentials(): IMatrixClientCreds;

    /**
     * If we've registered a user ID we set this to the ID of the
     * user we've just registered. If they then go & log in, we
     * can send them to the welcome user (obviously this doesn't
     * guarentee they'll get a chat with the welcome user).
     *
     * @param {string} uid The user ID of the user we've just registered
     */
    setJustRegisteredUserId(uid: string): void;

    /**
     * Returns true if the current user has just been registered by this
     * client as determined by setJustRegisteredUserId()
     *
     * @returns {bool} True if user has just been registered
     */
    currentUserIsJustRegistered(): boolean;

    /**
     * If the current user has been registered by this device then this
     * returns a boolean of whether it was within the last N hours given.
     */
    userRegisteredWithinLastHours(hours: number): boolean;

    /**
     * Replace this MatrixClientPeg's client with a client instance that has
     * homeserver / identity server URLs and active credentials
     *
     * @param {IMatrixClientCreds} creds The new credentials to use.
     */
    replaceUsingCreds(creds: IMatrixClientCreds): void;
}

/**
 * Wrapper object for handling the js-sdk Matrix Client object in the react-sdk
 * Handles the creation/initialisation of client objects.
 * This module provides a singleton instance of this class so the 'current'
 * Matrix Client object is available easily.
 */
class _MatrixClientPeg implements IMatrixClientPeg {
    // These are the default options used when when the
    // client is started in 'start'. These can be altered
    // at any time up to after the 'will_start_client'
    // event is finished processing.
    public opts: IOpts = {
        initialSyncLimit: 20,
    };

    private matrixClient: MatrixClient = null;
    private justRegisteredUserId: string;

    // the credentials used to init the current client object.
    // used if we tear it down & recreate it with a different store
    private currentClientCreds: IMatrixClientCreds;

    constructor() {
    }

    public setIndexedDbWorkerScript(script: string): void {
        createMatrixClient.indexedDbWorkerScript = script;
    }

    public get(): MatrixClient {
        return this.matrixClient;
    }

    public unset(): void {
        this.matrixClient = null;

        MatrixActionCreators.stop();
    }

    public setJustRegisteredUserId(uid: string): void {
        this.justRegisteredUserId = uid;
        if (uid) {
            window.localStorage.setItem("mx_registration_time", String(new Date().getTime()));
        }
    }

    public currentUserIsJustRegistered(): boolean {
        return (
            this.matrixClient &&
            this.matrixClient.credentials.userId === this.justRegisteredUserId
        );
    }

    public userRegisteredWithinLastHours(hours: number): boolean {
        try {
            const date = new Date(window.localStorage.getItem("mx_registration_time"));
            return ((new Date().getTime() - date.getTime()) / 36e5) <= hours;
        } catch (e) {
            return false;
        }
    }

    public replaceUsingCreds(creds: IMatrixClientCreds): void {
        this.currentClientCreds = creds;
        this.createClient(creds);
    }

    public async assign(): Promise<any> {
        for (const dbType of ['indexeddb', 'memory']) {
            try {
                const promise = this.matrixClient.store.startup();
                console.log("MatrixClientPeg: waiting for MatrixClient store to initialise");
                await promise;
                break;
            } catch (err) {
                if (dbType === 'indexeddb') {
                    console.error('Error starting matrixclient store - falling back to memory store', err);
                    this.matrixClient.store = new MemoryStore({
                        localStorage: localStorage,
                    });
                } else {
                    console.error('Failed to start memory store!', err);
                    throw err;
                }
            }
        }

        StorageManager.trackStores(this.matrixClient);

        // try to initialise e2e on the new client
        try {
            // check that we have a version of the js-sdk which includes initCrypto
            if (!SettingsStore.getValue("lowBandwidth") && this.matrixClient.initCrypto) {
                await this.matrixClient.initCrypto();
                this.matrixClient.setCryptoTrustCrossSignedDevices(
                    !SettingsStore.getValue('e2ee.manuallyVerifyAllSessions'),
                );
                await tryToUnlockSecretStorageWithDehydrationKey(this.matrixClient);
                StorageManager.setCryptoInitialised(true);
            }
        } catch (e) {
            if (e && e.name === 'InvalidCryptoStoreError') {
                // The js-sdk found a crypto DB too new for it to use
                const CryptoStoreTooNewDialog =
                    sdk.getComponent("views.dialogs.CryptoStoreTooNewDialog");
                Modal.createDialog(CryptoStoreTooNewDialog);
            }
            // this can happen for a number of reasons, the most likely being
            // that the olm library was missing. It's not fatal.
            console.warn("Unable to initialise e2e", e);
        }

        const opts = utils.deepCopy(this.opts);
        // the react sdk doesn't work without this, so don't allow
        opts.pendingEventOrdering = "detached";
        opts.lazyLoadMembers = true;
        opts.clientWellKnownPollPeriod = 2 * 60 * 60; // 2 hours

        // Connect the matrix client to the dispatcher and setting handlers
        MatrixActionCreators.start(this.matrixClient);
        MatrixClientBackedSettingsHandler.matrixClient = this.matrixClient;

        return opts;
    }

    public async start(): Promise<any> {
        const opts = await this.assign();

        console.log(`MatrixClientPeg: really starting MatrixClient`);
        await this.get().startClient(opts);
        console.log(`MatrixClientPeg: MatrixClient started`);
    }

    public getCredentials(): IMatrixClientCreds {
        return {
            homeserverUrl: this.matrixClient.baseUrl,
            identityServerUrl: this.matrixClient.idBaseUrl,
            userId: this.matrixClient.credentials.userId,
            deviceId: this.matrixClient.getDeviceId(),
            accessToken: this.matrixClient.getAccessToken(),
            guest: this.matrixClient.isGuest(),
        };
    }

    public getHomeserverName(): string {
        const matches = /^@[^:]+:(.+)$/.exec(this.matrixClient.credentials.userId);
        if (matches === null || matches.length < 1) {
            throw new Error("Failed to derive homeserver name from user ID!");
        }
        return matches[1];
    }

    private createClient(creds: IMatrixClientCreds): void {
        const opts: ICreateClientOpts = {
            baseUrl: creds.homeserverUrl,
            idBaseUrl: creds.identityServerUrl,
            accessToken: creds.accessToken,
            userId: creds.userId,
            deviceId: creds.deviceId,
            pickleKey: creds.pickleKey,
            timelineSupport: true,
            forceTURN: !SettingsStore.getValue('webRtcAllowPeerToPeer'),
            fallbackICEServerAllowed: !!SettingsStore.getValue('fallbackICEServerAllowed'),
            // Gather up to 20 ICE candidates when a call arrives: this should be more than we'd
            // ever normally need, so effectively this should make all the gathering happen when
            // the call arrives.
            iceCandidatePoolSize: 20,
            verificationMethods: [
                verificationMethods.SAS,
                SHOW_QR_CODE_METHOD,
                verificationMethods.RECIPROCATE_QR_CODE,
            ],
            unstableClientRelationAggregation: true,
            identityServer: new IdentityAuthClient(),
            cryptoCallbacks: {},
        };

        // These are always installed regardless of the labs flag so that
        // cross-signing features can toggle on without reloading and also be
        // accessed immediately after login.
        Object.assign(opts.cryptoCallbacks, crossSigningCallbacks);
        if (SecurityCustomisations.getDehydrationKey) {
            opts.cryptoCallbacks.getDehydrationKey =
                SecurityCustomisations.getDehydrationKey;
        }

        this.matrixClient = createMatrixClient(opts);

        // we're going to add eventlisteners for each matrix event tile, so the
        // potential number of event listeners is quite high.
        this.matrixClient.setMaxListeners(500);

        this.matrixClient.setGuest(Boolean(creds.guest));

        const notifTimelineSet = new EventTimelineSet(null, {
            timelineSupport: true,
        });
        // XXX: what is our initial pagination token?! it somehow needs to be synchronised with /sync.
        notifTimelineSet.getLiveTimeline().setPaginationToken("", EventTimeline.BACKWARDS);
        this.matrixClient.setNotifTimelineSet(notifTimelineSet);
    }
}

if (!window.mxMatrixClientPeg) {
    window.mxMatrixClientPeg = new _MatrixClientPeg();
}

export const MatrixClientPeg = window.mxMatrixClientPeg;
