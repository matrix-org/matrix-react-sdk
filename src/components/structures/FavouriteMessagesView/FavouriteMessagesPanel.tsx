/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { MatrixClient, MatrixEvent } from 'matrix-js-sdk/src/matrix';
import React, { useRef } from 'react';

import { _t } from '../../../languageHandler';
import ResizeNotifier from '../../../utils/ResizeNotifier';
import ScrollPanel from '../ScrollPanel';
import FavouriteMessagesHeader from './FavouriteMessagesHeader';
import FavouriteMessagesTilesList from './FavouriteMessagesTilesList';

interface IProps{
    favouriteMessageEvents?: MatrixEvent[];
    resizeNotifier?: ResizeNotifier;
    searchQuery?: string;
    handleSearchQuery?: (query: string) => void;
    cli?: MatrixClient;
}

const FavouriteMessagesPanel = (props: IProps) => {
    const favouriteMessagesPanelRef = useRef<ScrollPanel>();
    let favouriteMessagesPanel;

    if (props.favouriteMessageEvents?.length === 0) {
        favouriteMessagesPanel = (
            <>
                <FavouriteMessagesHeader handleSearchQuery={props.handleSearchQuery} />
                <h2 className="mx_FavouriteMessages_emptyMarker">{ _t("No Saved Messages") }</h2>
            </>
        );
    } else {
        favouriteMessagesPanel = (
            <>
                <FavouriteMessagesHeader handleSearchQuery={props.handleSearchQuery} />
                <ScrollPanel
                    ref={favouriteMessagesPanelRef}
                    className="mx_RoomView_searchResultsPanel mx_FavouriteMessages_scrollPanel"
                    resizeNotifier={props.resizeNotifier}
                >
                    <FavouriteMessagesTilesList cli={props.cli} favouriteMessageEvents={props.favouriteMessageEvents} favouriteMessagesPanelRef={favouriteMessagesPanelRef} searchQuery={props.searchQuery} />
                </ScrollPanel>
            </>
        );
    }

    return <>{ favouriteMessagesPanel }</>;
};
export default FavouriteMessagesPanel;

