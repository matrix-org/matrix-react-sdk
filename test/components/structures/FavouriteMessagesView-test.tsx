/*
Copyright 2022 The Matrix.org Foundation C.I.C.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
// eslint-disable-next-line deprecate/import
import { mount } from "enzyme";
import { mocked, MockedObject } from "jest-mock";
import { EventType, MatrixClient, MatrixEvent, MsgType } from "matrix-js-sdk/src/matrix";

import _FavouriteMessagesView from "../../../src/components/structures/FavouriteMessagesView/FavouriteMessagesView";
import { stubClient, mockPlatformPeg, unmockPlatformPeg, wrapInMatrixClientContext } from "../../test-utils";
import { MatrixClientPeg } from "../../../src/MatrixClientPeg";
import FavouriteMessagesPanel from "../../../src/components/structures/FavouriteMessagesView/FavouriteMessagesPanel";
import SettingsStore from "../../../src/settings/SettingsStore";

const FavouriteMessagesView = wrapInMatrixClientContext(_FavouriteMessagesView);

describe("FavouriteMessagesView", () => {
    let cli: MockedObject<MatrixClient>;
    // let room: Room;
    const userId = '@alice:server.org';
    const roomId = '!room:server.org';
    const alicesFavouriteMessageEvent = new MatrixEvent({
        type: EventType.RoomMessage,
        sender: userId,
        room_id: roomId,
        content: {
            msgtype: MsgType.Text,
            body: 'i am alice',
        },
        event_id: "$alices_message",
    });

    const bobsFavouriteMessageEvent = new MatrixEvent({
        type: EventType.RoomMessage,
        sender: '@bob:server.org',
        room_id: roomId,
        content: {
            msgtype: MsgType.Text,
            body: 'i am bob',
        },
        event_id: "$bobs_message",
    });

    beforeEach(async () => {
        mockPlatformPeg({ reload: () => {} });
        stubClient();
        cli = mocked(MatrixClientPeg.get());
    });

    afterEach(async () => {
        unmockPlatformPeg();
        jest.restoreAllMocks();
    });

    describe('favourite_messages feature when enabled', () => {
        beforeEach(() => {
            jest.spyOn(SettingsStore, 'getValue')
                .mockImplementation(setting => setting === 'feature_favourite_messages');
        });

        it('renders <FavouriteMessagesView /> correctly', () => {
            const view = mount(<FavouriteMessagesView />);
            expect(view.html()).toMatchSnapshot();
        });

        it('renders <FavouriteMessagesPanel /> component with empty or default props correctly', () => {
            const props = {
                favouriteMessageEvents: [],
                handleSearchQuery: jest.fn(),
                cli,
            };
            const view = mount(<FavouriteMessagesPanel {...props} />);
            expect(view.prop('favouriteMessageEvents')).toHaveLength(0);
            expect(view.contains("No Saved Messages")).toBeTruthy();
        });

        it('renders starred messages correctly for a single event', () => {
            const props = {
                favouriteMessageEvents: [bobsFavouriteMessageEvent],
                handleSearchQuery: jest.fn(),
                cli,
            };
            const view = mount(<FavouriteMessagesPanel {...props} />);

            expect(view.find('.mx_EventTile_body').text()).toEqual("i am bob");
        });

        it('renders starred messages correctly for multiple single event', () => {
            const props = {
                favouriteMessageEvents: [alicesFavouriteMessageEvent, bobsFavouriteMessageEvent],
                handleSearchQuery: jest.fn(),
                cli,
            };
            const view = mount(<FavouriteMessagesPanel {...props} />);
            //for alice
            expect(view.find("li[data-event-id='$alices_message']")).toBeDefined();
            expect(view.find("li[data-event-id='$alices_message']").contains("i am alice")).toBeTruthy();

            //for bob
            expect(view.find("li[data-event-id='$bobs_message']")).toBeDefined();
            expect(view.find("li[data-event-id='$bobs_message']").contains("i am bob")).toBeTruthy();
        });
    });
});
