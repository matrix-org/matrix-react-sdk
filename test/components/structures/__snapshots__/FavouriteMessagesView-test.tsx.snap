// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavouriteMessagesView when favourite_messages feature is enabled renders <FavouriteMessagesView /> correctly 1`] = `"<div class=\\"mx_FavMessagesHeader\\"><div class=\\"mx_FavMessagesHeader_Wrapper\\"><div class=\\"mx_FavMessagesHeader_Wrapper--left\\"><span class=\\"mx_BaseAvatar\\" role=\\"presentation\\"><span class=\\"mx_BaseAvatar_initial\\" aria-hidden=\\"true\\" style=\\"font-size: 16.900000000000002px; width: 26px; line-height: 26px;\\">F</span><img class=\\"mx_BaseAvatar_image\\" src=\\"data:image/png;base64,00\\" alt=\\"\\" style=\\"width: 26px; height: 26px;\\" aria-hidden=\\"true\\"></span><span>Favourite Messages</span></div><div class=\\"mx_FavMessagesHeader_Wrapper--right\\"><div aria-label=\\"Search\\" role=\\"button\\" tabindex=\\"0\\" class=\\"mx_AccessibleButton mx_RoomHeader_button mx_RoomHeader_searchButton\\"></div><div aria-label=\\"Reorder\\" role=\\"button\\" tabindex=\\"0\\" class=\\"mx_AccessibleButton mx_RoomHeader_button mx_FavMessagesHeader_sortButton\\"></div><div aria-label=\\"Clear\\" role=\\"button\\" tabindex=\\"0\\" class=\\"mx_AccessibleButton mx_RoomHeader_button mx_FavMessagesHeader_clearAllButton\\"></div></div></div></div><div class=\\"mx_AutoHideScrollbar mx_ScrollPanel mx_RoomView_searchResultsPanel mx_FavouriteMessages_scrollPanel\\" tabindex=\\"-1\\"><div class=\\"mx_RoomView_messageListWrapper\\"><ol class=\\"mx_RoomView_MessageList\\" aria-live=\\"polite\\" style=\\"height: 400px;\\"><div class=\\"mx_Spinner\\"><div class=\\"mx_Spinner_icon\\" style=\\"width: 32px; height: 32px;\\" aria-label=\\"Loading...\\" role=\\"progressbar\\"></div></div></ol></div></div>"`;
